### PR TITLE
Save IP addresses

### PR DESCRIPTION
**Pull request by Gaweringo**
This PR aims to implement issue #27

Showcase of the selecting last used IPs

A new IP is saved into the list of last used IPs when connecting to a new printer. If the IP is already in the list it is moved to the first place.

This also auto-fills and auto-selects the IP-Input field. Thus the user only needs to hit enter after the application starts if they want to connect to their last used printer.
This is not perfect, as the input field doesn't reflect its selected state in its color. I am not sure how I could change this.

When an IP from the last used list is selected it gets put into the IP Input field which is selected again.
It might be better if clicking the IP skipped the input field and connected to the printer with just one click.

Currently, the IPs are saved using localStorage which might make it complicated if more information (like the name) about a printer should be added to the list of saved IPs. For this, saving this data in a JSON file would probably be better.

### CODE REVIEW

**Comment by andycb** [comment]
Acknowledging this is on my radar. Thanks for adding, I think this is the number 1 most requested feature. 

I'll do a full review when I get a moment.

A quick thought, though: The IP history box as it stands will grow boundlessly, right? So if the user has a dynamic IP on their printer, we'll just be storing a load of random addresses. It feels like there should be a way to remove them. I imagine we'll need to use a similar UI to show the printer(s) we find from the UDP auto discovery, so I wonder if an auto-complete dropdown might work better for the history?

**Comment by Gaweringo** [comment]
Autocomplete sounds good, probably the better choice.
And for deleting. Maybe if we add settings this could be put there with a setting to turn the saving off completely.
Settings page with a list to delete saved IPs

**Comment by andycb** [comment]
Sounds good. 

I don't think there's a need for to go too complex with settings. So we probably wont need a full edit saved IPs experience, a clear button is probably sufficient (and much easier to test when releasing updates 😉)

File: src/electronApp/angularApp/components/connect-form/connect-form.component.ts [hunk 1]
```diff
@@ -1,35 +1,56 @@
-import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
-import { PrinterService } from '../../services/printerService';
-import { ErrorLogger } from '../../../core/errorLogger';
-import { FormControl, FormGroupDirective, NgForm, Validators} from '@angular/forms';
-import { ErrorStateMatcher} from '@angular/material/core';
-import { Router, ActivatedRoute } from '@angular/router';
+import {
```
Comment (andycb) [inline src/electronApp/angularApp/components/connect-form/connect-form.component.ts:1]: Can you update this, please, with your new formatter settings so there are fewer unrelated changes :)

File: src/electronApp/angularApp/components/connect-form/connect-form.component.ts [hunk 2]
```diff
@@ -72,9 +104,19 @@ export class ConnectFormComponent implements OnInit {
...
     });
+    this.ips = DataSaver.GetSavedIPs();
+    let lastIP = this.ips[0];
+    this.PrinterAddress.setValue(lastIP);
+  }
+
+  /**
+   * Invoked when the Angular component finished rendering.
+   */
+  ngAfterViewInit(): void {
```
Comment (andycb) [inline src/electronApp/angularApp/components/connect-form/connect-form.component.ts:118]: How come this is needed?

File: src/electronApp/angularApp/components/connect-form/connect-form.component.html [hunk 3]
```diff
@@ -2,21 +2,33 @@
   <h1>Hello!</h1>
 
   <h2 class="instructions">Enter printer's IP address to continue</h2>
-  <span *ngIf="isError" onkeypress="OnkeyPress" class="error warningColor">Failed to connect to printer, is the address correct?</span>
+  <span *ngIf="isError"
+        onkeypress="OnkeyPress"
```
Comment (andycb) [inline src/electronApp/angularApp/components/connect-form/connect-form.component.html:6]: Not your change, it was already here, but I don't think this click handler actually does anything

File: src/electronApp/core/dataSaver.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ */
+export class DataSaver {
+  static readonly lastIPsKey = "lastIPs";
+
+  /**
+   * Saves the IP-Address into the list of last used IPs
+   * @param ip The ip address to be saved
+   */
+  public static SaveLastIP(ip: string): void {
+    let LastIPs = localStorage.getItem(this.lastIPsKey);
```
Comment (andycb) [inline src/electronApp/core/dataSaver.ts:12]: minor: Can you use lowercase naming for locals, please

File: src/electronApp/core/dataSaver.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    }
+  }
+
+  /**
+   * Swaps the elements at old_index and new_index
+   * @param arr Array in which the elements are being moved
+   * @param old_index The index of the element to be moved
+   * @param new_index The index of the position the element will be moved to
+   */
+  static array_move(arr, old_index, new_index): void {
```
Comment (andycb) [inline src/electronApp/core/dataSaver.ts:40]: Guessing this was borrowed from here: https://stackoverflow.com/questions/5306680/move-an-array-element-from-one-array-position-to-another

It seems a lite more complex than it needs to be for what it is doing. I don't think it needs to deal with the case of the move index being off the end of the array, indeed this would be an error for the UI, so it would have been simpler to support move & validate the input is valid.

File: src/electronApp/core/dataSaver.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ * Save data locally
+ */
+export class DataSaver {
+  static readonly lastIPsKey = "lastIPs";
+
+  /**
+   * Saves the IP-Address into the list of last used IPs
+   * @param ip The ip address to be saved
+   */
+  public static SaveLastIP(ip: string): void {
```
Comment (andycb) [inline src/electronApp/core/dataSaver.ts:11]: I think we should probably only store the last x IPs to avid this having boundless growth.

**Review by andycb — CHANGES_REQUESTED** [review]
Thank again for doing this,.

I've made some comments on the code, as we spoke about before I think it would be best to have some kind of dropdown / auto complete rather than a box of IPs on the connect screen, too.

File: src/electronApp/angularApp/components/connect-form/connect-form.component.ts (hunk 2 above)
Comment (Gaweringo) [reply src/electronApp/angularApp/components/connect-form/connect-form.component.ts:97]: This is needed to focus on the input field. If I just use the autofocus tag on the HTML element it only focuses the first time the connect screen is shown (when the application is opened). And if the click is set onInit the UI doesn't update accordingly.
But I'm not really happy with the current implementation as the input looks like this.
Input field autofocus (looks like it's not active)
and not active like when I click with my mouse:
Clicked input field (looks active)

I think I should implement this functionality with `autofocus` but I don't know why it only works on the first try. Maybe it's the way how the routing works.

This is how autofocus currently behaves:
The input field is focused on the first time it's shown, but not after that

Comment (Gaweringo) [reply src/electronApp/angularApp/components/connect-form/connect-form.component.ts:97]: Looks like the IP list had something to do with this autofocus problem.

In this gif, I removed the IP list and added the autocomplete functionality according to [this example from Angular Material](https://stackblitz.com/angular/rmplyklknej?file=src%2Fapp%2Fautocomplete-auto-active-first-option-example.ts). The autofocus is done with `ngAfterViewInit()`. The input field is now correctly highlighted, but the autocomplete dropdown doesn't show up until the field is clicked or written to. The `autofocus` tag on the HTML element would show this, but again only the first time.
The input field is auto-focused and displayed correctly, but the autocomplete panel doesn't show up

And somehow with this change, the IPs aren't saved between startups, when running with `npm run electron` or after that with `electron .` but if I build the app and install it, the IPs are saved.

Comment (Gaweringo) [reply src/electronApp/angularApp/components/connect-form/connect-form.component.ts]: Fixed the autofocus with 776de02.
Based on [this answer](https://stackoverflow.com/questions/41873893/angular2-autofocus-input-element/42745350#42745350) from StackOverflow.

File: src/electronApp/core/dataSaver.ts (hunk 6 above)
Comment (Gaweringo) [reply src/electronApp/core/dataSaver.ts:19]: I've for now limited the max IPs to 5 with 742767d

File: src/electronApp/angularApp/components/connect-form/connect-form.component.html (hunk 3 above)
Comment (Gaweringo) [reply src/electronApp/angularApp/components/connect-form/connect-form.component.html]: Yeah doesn't look like it, I removed it 18dafab

**Comment by andycb** [comment]
Hey, I see you've been pushing a few changes to this. There's a couple of comments still open, are you still working on this or are you ready for a review?

**Comment by Gaweringo** [comment]
I'm done with all the suggestions you gave, so it's ready for review. I didn't comment on everything as I thought it might spam your inbox.
Should have probably sent a comment saying I'm done with all the suggestions.

**Review by andycb — COMMENTED** [review]
Cool, No worries. I'll take a look, I was taking a couple of weeks away from GitHub to focus on other things anyway :)

**Comment by Gaweringo** [comment]
For some reasons I can't reply directly to [this review](https://github.com/andycb/AdventurerClientJS/pull/49#discussion_r584274270) but I personally really like the workflow of opening the App, hitting enter and being connected to the last printer I used, no need to touch the mouse there.
And about the concern with the connect button being hidden behind the autocomplete, if you use the App for the first time, there aren't any IPs saved, so no autocomplete is hiding the connect button.

One compromise might be to autofill the input with the last used ip (no autocomplete shown) with everything selected. This way the connect button is visible on every opening and you can just hit enter to connect, or type something to get the autocomplete.

And about the saved IPs during debugging, I don't know what happened there, but when I tried it again yesterday it worked ¯\_(ツ)_/¯.

**Comment by Gaweringo** [comment]
Changed it.
It now works like this:
The app is opened and 192.168.1.21 is auto-filled into the input field (no autocomplete is shown). By pressing enter the app connects to the specified IP. After disconnecting, the same IP as last used is auto-filled again, but by typing a new IP the autocomplete shows up and can be used.
